gax/src/test/java/com/google/api/gax/rpc/ClientContextTest.java

```diff
@@ -597,4 +605,116 @@ public void testUserAgentConcat() throws Exception {
+            false);
+    String endpointSelected =
+        ClientContext.getEndpoint(endpoint, mtlsEndpoint, switchToMtlsEndpointAllowed, provider);
+    assertEquals(mtlsEndpoint, endpointSelected);
+  }
+
+  @Test
+  public void testEndpointNotOverridable() throws IOException {
+    // Test the case that switching to mTLS endpoint is not allowed so the original endpoint is
+    // selected.
+    boolean switchToMtlsEndpointAllowed = false;
+    MtlsProvider provider =
+        new FakeMtlsProvider(
+            true,
+            MtlsEndpointUsagePolicy.AUTO,
+            FakeMtlsProvider.createTestMtlsKeyStore(),
+            "",
+            false);
+    String endpointSelected =
+        ClientContext.getEndpoint(endpoint, mtlsEndpoint, switchToMtlsEndpointAllowed, provider);
+    assertEquals(endpoint, endpointSelected);
+  }
+
+  @Test
+  public void testNoClientCertificate() throws IOException {
+    // Test the case that client certificates doesn't exists so the original endpoint is selected.
+    boolean switchToMtlsEndpointAllowed = true;
+    MtlsProvider provider =
+        new FakeMtlsProvider(true, MtlsEndpointUsagePolicy.AUTO, null, "", false);
+    String endpointSelected =
+        ClientContext.getEndpoint(endpoint, mtlsEndpoint, switchToMtlsEndpointAllowed, provider);
+    assertEquals(endpoint, endpointSelected);
+  }
+
+  @Test
+  public void testAlwaysUseMtlsEndpoint() throws IOException {
+    // Test the case that mTLS endpoint is always used.
+    boolean switchToMtlsEndpointAllowed = true;
+    MtlsProvider provider =
+        new FakeMtlsProvider(false, MtlsEndpointUsagePolicy.ALWAYS, null, "", false);
+    String endpointSelected =
+        ClientContext.getEndpoint(endpoint, mtlsEndpoint, switchToMtlsEndpointAllowed, provider);
+    assertEquals(mtlsEndpoint, endpointSelected);
+  }
+
+  @Test
+  public void testNeverUseMtlsEndpoint() throws IOException {
+    // Test the case that mTLS endpoint is never used.
+    boolean switchToMtlsEndpointAllowed = true;
+    MtlsProvider provider =
+        new FakeMtlsProvider(
+            true,
+            MtlsEndpointUsagePolicy.NEVER,
+            FakeMtlsProvider.createTestMtlsKeyStore(),
+            "",
+            false);
+    String endpointSelected =
+        ClientContext.getEndpoint(endpoint, mtlsEndpoint, switchToMtlsEndpointAllowed, provider);
+    assertEquals(endpoint, endpointSelected);
+  }
+
+  @Test
+  public void testGetKeyStoreThrows() throws IOException {
+    // Test the case that getKeyStore throws exceptions.
+    try {
+      boolean switchToMtlsEndpointAllowed = true;
+      MtlsProvider provider =
+          new FakeMtlsProvider(true, MtlsEndpointUsagePolicy.AUTO, null, "", true);
+      ClientContext.getEndpoint(endpoint, mtlsEndpoint, switchToMtlsEndpointAllowed, provider);
+      fail("should throw an exception");
+    } catch (IOException e) {
+      assertTrue("expected to fail", e.getMessage().contains("getKeyStore throws exception"));
```

Comment:
the optional message argument should explain what went wrong if we got here and the assert failed, not what happened if this line isn't reached

Reply:
Done.